### PR TITLE
[sival/aes] Add AES Interrupt Encryption tests

### PR DESCRIPTION
**Pull request by rprakas-gsc**
@moidx  Checking in aes_interrupt_encryption_test.c code and modified the BUILD to include this test.

### CODE REVIEW

File: sw/device/tests/aes_interrupt_encryption_test.c [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+                                    (size_t)kAesNumBlocks_t2));
+
+  CHECK_DIF_OK(dif_aes_load_data(&aes2, plain_text2[0]));
+  LOG_INFO("Started AES trial Encrypt");
+  CHECK_DIF_OK(dif_aes_end(&aes2));
+
+  /** Not checking successful encryption here since its a trial transaction
+   *  Trial encryption ends here
+   *  Resume the Original Encryption below
+   */
```
Comment (luismarques) [inline sw/device/tests/aes_interrupt_encryption_test.c:182]: Nit: why is this using documentation-style comments?

File: sw/device/tests/BUILD [hunk 2]
```diff
@@ -6279,3 +6279,24 @@ opentitan_test(
...
     ],
 )
+
+opentitan_test(
+    name = "aes_interrupt_encryption_test",
+    srcs = ["aes_interrupt_encryption_test.c"],
+    exec_env = {
+        #         "//hw/top_earlgrey:sim_verilator": None,
+        "//hw/top_earlgrey:fpga_cw340_sival": None,
+        #         "//hw/top_earlgrey:fpga_cw340_rom_with_fake_keys": None,
```
Comment (luismarques) [inline sw/device/tests/BUILD:6289]: Please clean up the unsupported execution environments. Though perhaps you'd be able to add back verilator, etc?

File: sw/device/tests/aes_interrupt_encryption_test.c [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+  } while (!ready);
+
+  // Disable the AES unit to no longer automatically start encryption/decryption
+
+  transaction.manual_operation = kDifAesManualOperationManual;
+
+  CHECK_DIF_OK(dif_aes_trigger(&aes, kDifAesTriggerDataOutClear));
+  CHECK_DIF_OK(dif_aes_trigger(&aes, kDifAesTriggerKeyIvDataInClear));
+
+  LOG_INFO(" De-initialized Automatic operation ");
```
Comment (luismarques) [inline sw/device/tests/aes_interrupt_encryption_test.c:122]: Nit: why the extra whitespace in the string?

File: sw/device/tests/aes_interrupt_encryption_test.c [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+  memcpy(iv.iv, &iv_encrypt1, sizeof(iv.iv));
+  transaction.manual_operation = kDifAesManualOperationAuto;
+
+  CHECK_DIF_OK(dif_aes_start(&aes, &transaction, &key, &iv));
+
+  memcpy(plain_text[0].data, kAesModesPlainText_block34,
+         sizeof(kAesModesPlainText_block34));
+
+  CHECK_DIF_OK(dif_aes_get_status(&aes, kDifAesStatusInputReady, &ready));
+  LOG_INFO("Resuming Encrypt with the remaining 2 blocks of data ");
```
Comment (luismarques) [inline sw/device/tests/aes_interrupt_encryption_test.c:195]: Nit: ditto extraneous whitespace

**Review by luismarques — COMMENTED** [review]
Can you add review / commit comments explaining what overall scenario this test tests? I don't suppose there's a testplan testpoint to reference?

File: sw/device/tests/aes_interrupt_encryption_test.c [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+
+static const uint8_t kKeyShare1[] = {
+    0x0f, 0x1f, 0x2f, 0x3f, 0x4f, 0x5f, 0x6f, 0x7f, 0x8f, 0x9f, 0xaf,
+    0xbf, 0xcf, 0xdf, 0xef, 0xff, 0x0a, 0x1a, 0x2a, 0x3a, 0x4a, 0x5a,
+    0x6a, 0x7a, 0x8a, 0x9a, 0xaa, 0xba, 0xca, 0xda, 0xea, 0xfa,
+};
+
+OTTF_DEFINE_TEST_CONFIG();
+
+status_t execute_test(void) {
```
Comment (engdoreis) [inline sw/device/tests/aes_interrupt_encryption_test.c:48]: As this function returns `status_t`, you can replace all instances of `CHECK_DIF_OK` by `TRY`, which would make the code cleaner in my opinion.

File: sw/device/tests/aes_interrupt_encryption_test.c [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+
+enum {
+  kTestTimeout = (1000 * 1000),
+};
+
+// The mask share, used to mask kKey. Note that the masking should not be done
+// manually. Software is expected to get the key in two shares right from the
+// beginning.
+
+static const unsigned char kAesModesPlainText_block12[32] = {
```
Comment (engdoreis) [inline sw/device/tests/aes_interrupt_encryption_test.c:29]: The name convention for `const` variables.
```suggestion
static const unsigned char kAesModesPlainTextBlock12[32] = {
```
The same would apply throughout the file.

File: sw/device/tests/aes_interrupt_encryption_test.c (hunk 5 above)
Comment (rprakas-gsc) [reply sw/device/tests/aes_interrupt_encryption_test.c:48]: Ideally, I want the test code to report FAILURE & abort if the value doesn't get readout correctly , hence using `CHECK_DIF_OK`

**Comment by rprakas-gsc** [comment]
> Can you add review / commit comments explaining what overall scenario this test tests? I don't suppose there's a testplan testpoint to reference?
There is.  I have added the below lines in the new Commit 
Added AES Interrupt Encryption tests as per test plan identified here:[[…](https://opentitan.org/book/hw/top_earlgrey/data/chip_testplan.html?highlight=chip_sw_aes_interrupt_encryption#chip_sw_aes_interrupt_encryption)
  and modified BUILD

File: sw/device/tests/aes_interrupt_encryption_test.c (hunk 4 above)
Comment (rprakas-gsc) [reply sw/device/tests/aes_interrupt_encryption_test.c]: Fixed this in the updated aes_interrupt_encryption_Test.c code

File: sw/device/tests/BUILD (hunk 2 above)
Comment (rprakas-gsc) [reply sw/device/tests/BUILD]: Done!

File: sw/device/tests/aes_interrupt_encryption_test.c (hunk 1 above)
Comment (rprakas-gsc) [reply sw/device/tests/aes_interrupt_encryption_test.c]: Fixed these comments with regular-style

File: sw/device/tests/aes_interrupt_encryption_test.c (hunk 3 above)
Comment (rprakas-gsc) [reply sw/device/tests/aes_interrupt_encryption_test.c]: Fixed it - and amended the changes!

**Comment by moidx** [comment]
Hi @nasahlpa, can you take a look at the test implementation to ensure it matches the intended test point description? Thanks!

**Comment by nasahlpa** [comment]
> > Can you add review / commit comments explaining what overall scenario this test tests? I don't suppose there's a testplan testpoint to reference?
> > There is.  I have added the below lines in the new Commit
> > Added AES Interrupt Encryption tests as per test plan identified here:[[…](https://opentitan.org/book/hw/top_earlgrey/data/chip_testplan.html?highlight=chip_sw_aes_interrupt_encryption#chip_sw_aes_interrupt_encryption)
> > and modified BUILD

Could you please add the description to the git commit itself? Currently, no description is available in the git commit message. Furthermore, the contributing guideline also mentions that a [subject tag](https://opentitan.org/book/doc/contributing/detailed_contribution_guide/index.html#subject-tags) should be added to the git commit message.

**Comment by rprakas-gsc** [comment]
> > > Can you add review / commit comments explaining what overall scenario this test tests? I don't suppose there's a testplan testpoint to reference?
> > > There is.  I have added the below lines in the new Commit
> > > Added AES Interrupt Encryption tests as per test plan identified here:[[…](https://opentitan.org/book/hw/top_earlgrey/data/chip_testplan.html?highlight=chip_sw_aes_interrupt_encryption#chip_sw_aes_interrupt_encryption)
> > > and modified BUILD
> 
> Could you please add the description to the git commit itself? Currently, no description is available in the git commit message. Furthermore, the contributing guideline also mentions that a [subject tag](https://opentitan.org/book/doc/contributing/detailed_contribution_guide/index.html#subject-tags) should be added to the git commit message.

The git commit has restrictions on word limit, if I include the entire link, i will  run out of the world limit and won't be able to commit successfully.

**Comment by nasahlpa** [comment]
Understood. But you could add something like this:
```
This commit implements the chip_sw_aes_interrupt_encryption test that is defined in the chip testplan. The purpose of this test is to check, whether an AES encryption can be interrupted and restored.
```
For the header of the git commit message you could use:
```
[sival/aes] Add AES interrupt encryption test
```

**Comment by rprakas-gsc** [comment]
Thanks Pascal.  I will incorporate these changes in the commit message now!

**Comment by rprakas-gsc** [comment]
Just updated the PR with the commit messages as suggested!

**Comment by nasahlpa** [comment]
Thanks! Would you mind looking into the review comments I've added? Especially the question regarding the decryption should be addressed.

**Comment by moidx** [comment]
Hi @nasahlpa, I wasn't able to find your comments. Can you double check if they were sent? Thanks

**Review by nasahlpa — CHANGES_REQUESTED** [review]

**Comment by nasahlpa** [comment]
> Hi @nasahlpa, I wasn't able to find your comments. Can you double check if they were sent? Thanks

Oh sorry, now the comments should be public.

File: sw/device/tests/aes_interrupt_encryption_test.c (hunk 5 above)
Comment (engdoreis) [reply sw/device/tests/aes_interrupt_encryption_test.c:48]: The final result would be the same, `TRY` would return immediately and then the callee would abort by using `CHECK_STATUS_OK`.

File: sw/device/tests/BUILD
Comment (engdoreis) [inline sw/device/tests/BUILD:1]: Could you please re-order the commit message? The  `Signed-off` line should be after the description.

File: sw/device/tests/aes_interrupt_encryption_test.c
Comment (engdoreis) [inline sw/device/tests/aes_interrupt_encryption_test.c:1]: I suggest the following commit message:
[test, aes] Clean-up log message.

File: sw/device/tests/aes_interrupt_encryption_test.c [hunk 7]
```diff
@@ -0,0 +1,261 @@
...
+    //  Trial encryption ends here
+    //  Resume the Original Encryption below
+
+    // Using the iv read from `iv_encrypt1` for resuming the encryption of
+    // remaining 2 blocks.
+
+    memcpy(iv.iv, &iv_encrypt1, sizeof(iv.iv));
+    transaction.manual_operation = kDifAesManualOperationAuto;
+
+    // "Convert" key share byte arrays to `dif_aes_key_share_t`.
```
Comment (nasahlpa) [inline sw/device/tests/aes_interrupt_encryption_test.c:198]: I think this is not necessary as `key` should still contain the key used for the first two blocks?

**Review by nasahlpa — APPROVED** [review]
Thanks @rprakas-gsc for changing the code based on the feedback. Apart from a small nit this PR looks good to me.

Maybe @engdoreis could have another look.

**Review by engdoreis — APPROVED** [review]
Thanks for the changes.

**Comment by jwnrt** [comment]
@rprakas-gsc sorry we missed this in review, but would you mind filing a PR that assigns this test to the testpoint in the testplan please?

https://github.com/lowRISC/opentitan/blob/3a4d766b45e58af05dfccd81a861927724a4783f/hw/top_earlgrey/data/ip/chip_aes_testplan.hjson#L115

i.e.:
```
bazel: ["//sw/device/tests/aes_interrupt_encryption_test"]
```

Feel free to add me as a reviewer

**Comment by rprakas-gsc** [comment]
Thanks James! I have create a separate PR request (https://github.com/lowRISC/opentitan/pull/25501) for this.

**Pull request merged**